File: compiler/nodes/statements/assignment.go
package statements

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/patterns"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type AssignmentNode struct {
	Pattern database.Node
	Value   database.Node
	Facts   *database.Facts

	temporary database.Node
}

func (node *AssignmentNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseAssignmentStatement(parser *syntax.Parser, parseExpression ParseExpressionFunc) (*AssignmentNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := ParseComments(parser)
	if err != nil {
		return nil, err
	}

	pattern, err := patterns.ParsePattern(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("AssignOperator")
	if err != nil {
		return nil, err
	}

	parser.Commit("in this variable assignment")

	parser.ConsumeLineBreaks()

	value, err := parseExpression(parser)
	if err != nil {
		return nil, err
	}

	return &AssignmentNode{
		Pattern: pattern,
		Value:   value,
		Facts:   database.NewFacts(span()),
	}, nil
}

func (node *AssignmentNode) Visit(visitor *visit.Visitor) {
	visitor.AfterAllDefinitions(func() {
		// Try assigning to an existing constant if possible
		if pattern, ok := node.Pattern.(*patterns.VariablePatternNode); ok {
			if definitions, ok := visit.Peek[*visit.ConstantDefinition](visitor, pattern.Variable); ok {
				definition := definitions[len(definitions)-1]
				if !definition.Assigned {
					constantType := definition.Value

					visit.Defining(visitor, definition.Node, func() (*visit.ConstantDefinition, bool) {
						visitor.CurrentDefinition.WithinConstantValue = true

						visitor.Visit(node.Value)

						visitor.Constraint(typecheck.NewGroupConstraint(node.Value, constantType))

						definition.Assigned = true
						definition.Value = node.Value

						return nil, false
					})

					return
				}
			}
		}

		visitor.Visit(node.Value)

		node.temporary = &database.HiddenNode{
			Facts: database.NewFacts(database.GetSpanFact(node)),
		}

		visitor.Db.Register(node.temporary)

		visit.Matching(visitor, node.temporary, true, func() struct{} {
			visitor.Visit(node.Pattern)
			return struct{}{}
		})

		visitor.Constraint(typecheck.NewGroupConstraint(node.Pattern, node.Value))
	})
}

func (node *AssignmentNode) Codegen(c *codegen.Codegen) error {
	if node.temporary == nil {
		return nil // assigned to constant
	}

	span := database.GetSpanFact(node)

	c.WriteString(span, "var ")
	c.WriteNode(span, node.temporary)
	c.WriteString(span, ";")
	c.WriteLine()

	for _, temporary := range patterns.CollectTemporaries(node.Pattern) {
		if temporary == node.temporary {
			continue
		}

		c.WriteString(span, "var ")
		c.WriteNode(span, temporary)
		c.WriteString(span, ";")
		c.WriteLine()
	}

	c.WriteNode(span, node.temporary)
	c.WriteString(span, " = ")
	if err := c.Write(node.Value); err != nil {
		return err
	}
	c.WriteString(span, ";")
	c.WriteLine()

	c.WriteString(span, "if (true")
	if err := c.Write(node.Pattern); err != nil {
		return err
	}
	c.WriteString(span, `) {} else { throw new Error("unreachable"); }`)
	c.WriteLine()

	return nil
}

File: compiler/src/nodes/statements/assignment.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Fact, Node, NodeRef, Render},
    nodes::{Temporaries, VariablePatternNode, parse_comments, parse_expression, parse_pattern},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::{GroupConstraint, TypeConstraint},
    visit::{Definition, Visit, Visitor},
};

#[derive(Debug, Clone)]
pub struct ResolvedConstantAssignment;

impl Fact for ResolvedConstantAssignment {}

impl Render for ResolvedConstantAssignment {}

#[derive(Debug)]
pub struct AssignmentNode {
    pub pattern: NodeRef,
    pub value: NodeRef,
}

impl Node for AssignmentNode {
    fn is_hidden(&self) -> bool {
        true
    }
}

pub fn parse_assignment_statement(parser: &mut Parser<'_>) -> Result<AssignmentNode, ParseError> {
    let _ = parse_comments(parser)?;

    let pattern = parse_pattern(parser)?;

    parser.token(TokenKind::AssignOperator)?;
    parser.commit("in this variable assignment");
    parser.consume_line_breaks();

    let value = parse_expression(parser)?;

    Ok(AssignmentNode { pattern, value })
}

impl Visit for AssignmentNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visitor.constraint(TypeConstraint::new(node.clone(), visitor.unit_type()));

        let pattern = self.pattern.clone();
        let value = self.value.clone();
        let node = node.clone();
        visitor.after_all_definitions(move |visitor| {
            // Try assigning to an existing constant if possible
            if let Some(pattern_node) = pattern.downcast_ref::<VariablePatternNode>() {
                let found_constant = visitor.with(&pattern_node.variable, |visitor, definition| {
                    let Definition::Constant(definition) = definition else {
                        return None;
                    };

                    if definition.value.is_some() {
                        // TODO: Already assigned
                        return None;
                    }

                    let mut definition = definition.clone();
                    definition.value = Some(value.clone());

                    let definition_node = definition.node.clone();
                    visitor.defining(&definition_node, |visitor| {
                        visitor.within_constant_value(|visitor| {
                            visitor.visit(&value);
                            visitor.constraint(GroupConstraint::new(
                                value.clone(),
                                definition_node.clone(),
                            ));
                        });

                        definition.clone()
                    });

                    visitor.insert(
                        &node,
                        Temporaries {
                            has: Vec::new(),
                            inherit: Vec::new(),
                        },
                    );

                    visitor.insert(&node, ResolvedConstantAssignment);

                    Some(definition.into())
                });

                if found_constant {
                    return;
                }
            }

            visitor.visit(&value);
            visitor.edge(&value, &pattern, "value");

            visitor.matching(&value, false, true, |visitor| {
                visitor.current_match().root = Some(value.clone());
                visitor.current_match().arm = Some(pattern.clone());
                visitor.visit(&pattern);
            });

            visitor.constraint(GroupConstraint::new(pattern.clone(), value.clone()));
        });
    }
}

impl Codegen for AssignmentNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        if ctx.get::<ResolvedConstantAssignment>(node).is_some() {
            return Ok(());
        }

        ctx.codegen(&self.value)?;

        ctx.push_conditions();
        ctx.codegen(&self.pattern)?;
        let conditions = ctx.pop_conditions();

        ctx.instruction(ir::Instruction::If {
            node: None,
            branches: vec![(conditions, Vec::new(), None)],
            else_branch: None,
        });

        ctx.instruction(ir::Instruction::Value {
            node: node.clone(),
            value: ir::Value::Tuple(Vec::new()),
        });

        Ok(())
    }
}
